cmd/jaeger/internal/extension/jaegerstorage/extension.go

```diff
@@ -60,35 +62,62 @@ func newStorageExt(config *Config, otel component.TelemetrySettings) *storageExt
 	}
 }
 
-func (s *storageExt) Start(ctx context.Context, host component.Host) error {
-	for name, mem := range s.config.Memory {
-		if _, ok := s.factories[name]; ok {
-			return fmt.Errorf("duplicate memory storage name %s", name)
-		}
-		s.factories[name] = memory.NewFactoryWithConfig(
-			mem,
-			metrics.NullFactory,
-			s.logger.With(zap.String("storage_name", name)),
-		)
-	}
+type starter[Config any, Factory storage.Factory] struct {
+	ext         *storageExt
+	storageKind string
+	cfg         map[string]Config
+	builder     func(Config, metrics.Factory, *zap.Logger) (Factory, error)
+}
 
-	for name, b := range s.config.Badger {
-		if _, ok := s.factories[name]; ok {
-			return fmt.Errorf("duplicate badger storage name %s", name)
+func (s *starter[Config, Factory]) build(ctx context.Context, host component.Host) error {
```

Comment:
Internal state isn't persisted between method calls so all methods are effectively read-only; we could assert this by using value receivers instead of pointer receivers.

Reply:
true, but makes no difference in this context since the code runs exactly once. Don't want to reset the approval just for this fix :-)